src/GitTabExpansion.ps1

```diff
@@ -374,8 +380,15 @@ function GitTabExpansionInternal($lastBlock, $GitStatus = $null) {
             gitAddFiles $GitStatus $matches['files']
         }
 
-        # Handles git checkout -- <path>
-        "^checkout.* -- (?<files>\S*)$" {
+        # Handles git restore -s <ref> - must come before the next regex case
+        "^restore.* -s\s*(?<ref>\S*)$" {
+            gitBranches $matches['ref'] $true
+            gitTags $matches['ref']
+            break
+        }
+
+        # Handles git checkout -- <path> and git restore <path>
+        "^(?:checkout.* --|restore.*) (?<files>\S*)$" {
             gitCheckoutFiles $GitStatus $matches['files']
```

Comment:
I don't think this is correct; unless I'm reading things wrong it's missing support for `--staged`/`-S` . Even in that case,  files which qualify for a restore are more than just those that qualify for the "old" `checkout` -- `restore` is a bit more complex, enough-so that I pulled out a new function `gitRestoreFiles` for it. I don't have my laptop with me and unfortunately didn't push my latest updates to my fork, but best I can recall both with and without --staged could restore a *mixture* of Index and Working files. I originally had something very similar until I realized the files being auto-completed didn't match what git was suggesting. I'll check my local when I get into the office tomorrow.

Reply:
Unless I'm missing something, according to these docs - https://git-scm.com/docs/git-restore/2.23.0 - `-S` and `--staged` appear to behave like switch parameters (taking no value).  As far as files to restore the docs mention only `<pathspec>` which led me to believe it expects file paths.

Reply:
You are correct that those are switch parameters. You are also correct that `git restore` takes a `<pathspec>`, however what _qualifies_ as a valid pathspec seems to vary based on the presence/absence of the `--staged`/`-S`  switch. This is based on my somewhat extended use of the feature thus far:

- Presence of the `--staged` switch: 
  - Index.Modified
  - Index.Added
  - Index.Deleted
- Absence of the `--staged` switch: 
  - Working.Unmerged
  - Working.Modified
  - Working.Deleted

Maybe I'm missing something (or perhaps over-complicating it). However, I wasn't able to get the tab expansion of actual files in my tree to match the options git was presenting until I had limited to the above.

Reply:
Interesting.  I'm not sure we've done such a "stateful" tab-completion with this setup. I've done it using the built-in `Register-ArgumentCompleter` facilities.  But it should be doable.  Basically the difference is that with `--staged` you add in the "Added" files and without you add in the "Unmerged" files.  I think I can do that by capturing the `.*` in this case and examining it for the presence of `--staged|-S`.

Reply:
We do something similar when calling `gitDiffFiles` which also has a staged parameter. This is what I had come up with:

```
function script:gitRestoreFiles($GitStatus, $filter, $staged) {
    if ($staged) {
        gitFiles $filter (@($GitStatus.Index.Modified) +  @($GitStatus.Index.Added)+@($GitStatus.Index.Deleted))
    } else {
        gitFiles $filter (@($GitStatus.Working.Unmerged) + @($GitStatus.Working.Modified) + @($GitStatus.Working.Deleted))
    }
}
```
and

```
        "^restore(?:.* (?<staged>(?:--staged))|.*) (?<files>\S*)$" {
            gitRestoreFiles $GitStatus $matches['files'] $matches['staged']
        }
```

Though the regex might need tweaking, as I said I was having some difficulties with the -s and -S and I can't remember now if I fixed it completely.